### DISCUSSION TITLE
Issues With Running Sintering Model

### DISCUSSION THREAD

**Discussion by ibuhari86** [idaholab/moose#21523] [Q&A Modules: Phase field] [answered]
Hello, 

I am trying to run one of the input files from @SudiptaBiswas 2018 paper but I am running into issues. 

Since I am new to the MOOSE/GitHub environment, I started off with the first tutorial on app development, but since I am not trying to create a new application for a starter, I intended to run the input file "sintering_test1_isomob.i" from tests in Crow. I downloaded the crow as a zipped file, started with the creation of a new app named "crow" by following the first few steps of tutorial one (image 1 attached), and then pasting the content of the downloaded crow into the new crow I created. However, when I tried to execute the command `cd ~/projects/babbler
make -j4` it does not create the `crow-opt` file needed to run the app. I start to get fatal errors as seen in the second and third images. 

Any ideas what I might be missing? Or is there a better way to achieve this? Thank you. 
![Screenshot 2022-07-06 182603](https://user-images.githubusercontent.com/68480824/177654006-74ca6d6c-de19-41a5-b7c9-f16cb9285969.png)
![Screenshot 2022-07-06 182628](https://user-images.githubusercontent.com/68480824/177654014-9fb644b6-7c7c-41c5-943e-cd943a347c3e.png)
![Screenshot 2022-07-06 181117](https://user-images.githubusercontent.com/68480824/177653985-531b3851-3105-4f3e-b0b0-57a5c0033939.png)

**Comment by SudiptaBiswas** [comment, marked as answer]
Hi @ibuhari86, You will need to switch to `devel` branch to compile the application. Cloning the application instead of using the zip might work better.

**Reply by ibuhari86** [reply]
Hello @SudiptaBiswas, 

Thank you very much for your response. I tried cloning and when I executed the same command it seemed to be compiling alright at the start but exited with one of the fatal errors I encountered above. 
![Screenshot 2022-07-06 190428](https://user-images.githubusercontent.com/68480824/177657684-d2e21a83-d7e1-4d37-b32c-057bdb0198db.png)

**Reply by SudiptaBiswas** [reply]
You need to switch to devel branch. Do `git checkout devel` before compiling.

**Reply by ibuhari86** [reply]
I switched to devel branch, tried to compile and run into this (images 1 & 2). I had a similar issue when I was installing MOOSE for the first time, and I needed to delete `idaholab` from `mamba list` and reinstall the mamba moose environment and that solved the issue. `idaholab` is no linger an output of `cat ~/.condarc`, so I tried reinstalling the mamba package again but the error is persistent (image 3).
![Screenshot 2022-07-06 201200](https://user-images.githubusercontent.com/68480824/177664255-aaf09787-8d41-42b1-b491-7d883ed0a9b0.png)
![Screenshot 2022-07-06 201227](https://user-images.githubusercontent.com/68480824/177664258-6adfa3fb-a60d-41ea-ad1d-168c985f9ad7.png)
![Screenshot 2022-07-06 202923](https://user-images.githubusercontent.com/68480824/177664261-9a8c94f6-9fa4-40c4-81ff-c46b0db74004.png)

**Reply by SudiptaBiswas** [reply]
This indicates that the libmesh is not built. Try running the libmesh update script.

**Reply by ibuhari86** [reply]
Hi @SudiptaBiswas,

Are you talking about this script 
`cd ~/projects/moose`
`./scripts/update_and_rebuild_libmesh.sh`? 
![Screenshot 2022-07-07 005346](https://user-images.githubusercontent.com/68480824/177694811-0b153e74-0589-4756-9a94-7727a136da4a.png)
![Screenshot 2022-07-07 005419](https://user-images.githubusercontent.com/68480824/177694823-57a935e8-b29e-4db3-be76-03b2f85389fc.png)
If it is, then I run it and still could not compile.

**Reply by ibuhari86** [reply]
Now, when I run `make -j4` in the moose environment (rather than base), Error 127 disappears and another (Error 1) appears as shown. `./run_tests -j8` then resulted in FAILED tests.
![Screenshot 2022-07-07 020033](https://user-images.githubusercontent.com/68480824/177703672-943b7d89-d632-45d5-99a0-d6f778d84eef.png)
![Screenshot 2022-07-07 020124](https://user-images.githubusercontent.com/68480824/177703675-9da00308-012d-4f26-8009-8161d8104a9c.png)

**Reply by SudiptaBiswas** [reply]
This seems like a linking error. You will have to clean your directory before rebuilding the application. Try `make clobberall` first within the application and you might also need to perform `git clean -xfd` ( caution: this will delete any unchecked files, make sure to commit any changes you might have before doing this). Then try `make` again. It also might be a good idea to rebuild a moose module before trying the application.

**Reply by GiudGiud** [reply]
Hello

So for this last error, you are missing a function definition. 
You need to have a validParam function in all your materials, just like this (this is a kernel)
header:
```
class FVAdvection : public FVFluxKernel
{
public:
  static InputParameters validParams();
...
```
source:
```
InputParameters
FVAdvection::validParams()
{
  InputParameters params = FVFluxKernel::validParams();
  params.addClassDescription(
      "Residual contribution from advection operator for finite volume method.");
  params.addRequiredParam<RealVectorValue>("velocity", "Constant advection velocity");
  MooseEnum advected_interp_method("average upwind skewness-corrected", "upwind");

  params.addParam<MooseEnum>(
      "advected_interp_method",
      advected_interp_method,
      "The interpolation to use for the advected quantity. Options are "
      "'upwind', 'average' and 'skewness-corrected', with the default being 'upwind'.");
  return params;
}
```


The previous validParam syntax is deprecated, you need to update all of CrowApp to use the new one. 

Guillaume

**Reply by GiudGiud** [reply]
also you need to be more careful with the build instructions. You previously tried to build libmesh independently, but you are also trying to use the mamba install. Both methods are incompatible (except for expert users, they can be made to work together) and may lead to compiling errors.

**Reply by ibuhari86** [reply]
![Screenshot 2022-07-07 132011](https://user-images.githubusercontent.com/68480824/177832959-733a92ca-b9df-435f-bb0e-b6ca7213d7ff.png)
> This seems like a linking error. You will have to clean your directory before rebuilding the application. Try `make clobberall` first within the application and you might also need to perform `git clean -xfd` ( caution: this will delete any unchecked files, make sure to commit any changes you might have before doing this). Then try `make` again. It also might be a good idea to rebuild a moose module before trying the application.

Hi @SudiptaBiswas, by your last statement, did you mean reinstalling the MOOSE environment again from scratch? I did `make clobberall` and `git clean -xfd` but I am still getting Error 1.

**Reply by ibuhari86** [reply]
> Hello
> 
> So for this last error, you are missing a function definition. You need to have a validParam function in all your materials, just like this (this is a kernel) header:
> 
> ```
> class FVAdvection : public FVFluxKernel
> {
> public:
>   static InputParameters validParams();
> ...
> ```
> 
> source:
> 
> ```
> InputParameters
> FVAdvection::validParams()
> {
>   InputParameters params = FVFluxKernel::validParams();
>   params.addClassDescription(
>       "Residual contribution from advection operator for finite volume method.");
>   params.addRequiredParam<RealVectorValue>("velocity", "Constant advection velocity");
>   MooseEnum advected_interp_method("average upwind skewness-corrected", "upwind");
> 
>   params.addParam<MooseEnum>(
>       "advected_interp_method",
>       advected_interp_method,
>       "The interpolation to use for the advected quantity. Options are "
>       "'upwind', 'average' and 'skewness-corrected', with the default being 'upwind'.");
>   return params;
> }
> ```
> 
> The previous validParam syntax is deprecated, you need to update all of CrowApp to use the new one.
> 
> Guillaume

Hello GiudGiud, 

How can I get the latest version of CrowApp? The one I cloned was the devel version suggested by @SudiptaBiswas which I believe had the latest version? Thank you.

**Reply by SudiptaBiswas** [reply]
I am able to build the `devel` branch of CrowAPP fine on my mac. My environment might be a bit old though.

**Reply by SudiptaBiswas** [reply]
Although, I must admit that this application is old and not maintained regularly. It could be a while for me to update the syntax. @ibuhari86 It might be a good idea to try building the input file from MOOSE using `DerivativeParsedMaterial`.

**Reply by GiudGiud** [reply]
ok might not be that then. 
This kind of error is still usually symptomatic of a missing function implementation. Maybe some remains from a previous compilation are left and a function definition has changed since then if it builds fine for Sudipta

**Reply by ibuhari86** [reply]
> Although, I must admit that this application is old and not maintained regularly. It could be a while for me to update the syntax. @ibuhari86 It might be a good idea to try building the input file from MOOSE using `DerivativeParsedMaterial`.

Hi @SudiptaBiswas, I changed the 'type' in the Material block in 'sintering_test1_isomob.i' to `DerivativeParsedMaterial`, however, that did not solve the issue. Unless that is not the right way to go about it. Thanks.

**Reply by SudiptaBiswas** [reply]
That's not how you do it. You can find the documentation here,  https://mooseframework.inl.gov/moose/source/materials/DerivativeParsedMaterial.html. In general, I would recommend that you familiarize yourself with simple phase-field tests and examples available in MOOSE first. Since you are new to MOOSE, going through the tutorials, building input files, and solving some benchmark problems might be a good starting point.

**Reply by ibuhari86** [reply]
> That's not how you do it. You can find the documentation here, https://mooseframework.inl.gov/moose/source/materials/DerivativeParsedMaterial.html. In general, I would recommend that you familiarize yourself with simple phase-field tests and examples available in MOOSE first. Since you are new to MOOSE, going through the tutorials, building input files, and solving some benchmark problems might be a good starting point.

Thank you very much @SudiptaBiswas, I will proceed as advised. Thanks.

**Reply by ibuhari86** [reply]
Hi @SudiptaBiswas,

So when I switched to a different system (macOS), it compiled but the two tests still failed. 

![Capture1](https://user-images.githubusercontent.com/68480824/178359438-0ec971ec-1f35-49d9-9504-e2c1c49096d9.JPG)

**Reply by GiudGiud** [reply]
you have a missing implementation.
Basically a routine (computeQpEigenstrain) is declared in the header for computeEigenstrainTempl but never implemented in a source file. Possibly you should implement it for your derived class of computeEigenstrainTempl

**Reply by ibuhari86** [reply]
Thank you @GiudGiud.

**Reply by ibuhari86** [reply]
> you have a missing implementation. Basically a routine (computeQpEigenstrain) is declared in the header for computeEigenstrainTempl but never implemented in a source file. Possibly you should implement it for your derived class of computeEigenstrainTempl

Hi @GiudGiud,
Sorry to bother you again, I am new to this and have been going through several examples and tutorials in the documentation, however, I am not able to see how to implement this in the source file. Any further pointers would be appreciated. Thank you very much.

**Reply by GiudGiud** [reply]
can you push your current work to a git repo? 
I can try building it see what's missing exactly

**Reply by ibuhari86** [reply]
> can you push your current work to a git repo? I can try building it see what's missing exactly

Hi @GiudGiud, 

I am still using the devel branch of Crow which I was trying just to run the tests. I only cloned it to my local machine with the same name, but when I tried to push it I get errors. 

`nothing to commit, working tree clean
(base) buhari@me-Caradhras Crow % git remote add origin https://github.com/ibuhari86/Crow-cloned.git
error: remote origin already exists.
(base) buhari@me-Caradhras Crow % git remote -v
origin	https://github.com/SudiptaBiswas/Crow.git (fetch)
origin	https://github.com/SudiptaBiswas/Crow.git (push)
(base) buhari@me-Caradhras Crow % git push origin main
error: src refspec main does not match any
error: failed to push some refs to 'https://github.com/SudiptaBiswas/Crow.git'
(base) buhari@me-Caradhras Crow % 

I was thinking of using a different name for the cloned version but I was not sure if there could be name conflicts if the name "Crow" is called out in other files.

**Reply by GiudGiud** [reply]
Crow/devel builds fine for me.
It does have a few compilation warnings but that's it

**Reply by ibuhari86** [reply]
> Crow/devel builds fine for me.
> 
> It does have a few compilation warnings but that's it

And you don't get the failed test messages for the two tests I posted in the last image?

**Reply by GiudGiud** [reply]
oh I do my bad. 
Let me check if I can find out why

**Reply by ibuhari86** [reply]
> oh I do my bad. Let me check if I can find out why

@GiudGiud Sure, no problem. Thank you for the time.

**Reply by GiudGiud** [reply]
I noticed two things:
- in PFEigenStrainBaseMaterial.C 
you have computeStressFreeStrain() calling computeStressFreeStrain() directly. 
This will just recurse infinitely

- all your classes are using the deprecated and no longer supported validParam() templated syntax.
Please see this newsletter for more information
https://github.com/idaholab/moose/blob/8307e4f2cef9bae25e6babbf2b1869ea5c10065a/modules/doc/content/newsletter/2020_04.md

it should not even be building right now
Until this is fixed Crow will not be functional.

**Reply by ibuhari86** [reply]
> I noticed two things:
> 
> - in PFEigenStrainBaseMaterial.C 
> 
> you have computeStressFreeStrain() calling computeStressFreeStrain() directly. 
> 
> This will just recurse infinitely
> 
> 
> 
> - all your classes are using the deprecated and no longer supported validParam() templated syntax.
> 
> Please see this newsletter for more information
> 
> https://github.com/idaholab/moose/blob/8307e4f2cef9bae25e6babbf2b1869ea5c10065a/modules/doc/content/newsletter/2020_04.md
> 
> 
> 
> it should not even be building right now
> 
> Until this is fixed Crow will not be functional. 

Thanks @GiudGiud for your input. I'm assuming either the original developer or someone with permission can make those fixes, then?

**Reply by GiudGiud** [reply]
This code is open source, anyone needing to use Crow may make those fixes.

**Reply by ibuhari86** [reply]
Alright, thank you!

**Reply by ibuhari86** [reply]
@GiudGiud,

In the second problem, to update the `validParams()`, it made reference to making changes to the header file but it did not specify which. However, I noticed several header files in Crow, is there a particular one that needs to be updated? If so, is it in a specific folder? Thank you. 

Also, for the first issue, do you propose suppressing those lines of code that call `computeStressFreeStrain() directly` as a fix to the issue or there is more to it than that? Thank you.

**Reply by GiudGiud** [reply]
ALL the headers and sources that have the templated implementation of validParams need to be changed to 
so in Crow you potentially have up to 70 files to update I think

But it s a very easy change

I dont know what the implementation of comptueStressFreeStrain should be

**Reply by ibuhari86** [reply]
@GiudGiud, Thank you very much for your input.

**Reply by ibuhari86** [reply]
> @GiudGiud, I see. Anyway, thank you very much for your input.

**Reply by GiudGiud** [reply]
@fatimahdzr is hitting the same issue
Could you please collaborate to share the updated app with fixed validParams?
